### PR TITLE
feat(notifications): add notification action

### PR DESCRIPTION
**Pull request by UncleSamtoshi**
- add action for notification with options of OpenDeepLink and OpenExternalUrl

### CODE REVIEW

File: core/notifications/proto/notifications.proto [hunk 1]
```diff
@@ -238,17 +238,31 @@ message PriceChanged {
...
 }
 
+enum ActionType {
+  OPEN_DEEP_LINK = 0;
+  OPEN_EXTERNAL_URL = 1;
+}
+
+message Action {
+  ActionType type = 1;
+  oneof data {
```
Comment (bodymindarts) [inline core/notifications/proto/notifications.proto:260]: isn't type redundant when you use oneof?

Comment (thevaibhav-dixit) [reply core/notifications/proto/notifications.proto:254]: thanks.. fixed it.

File: core/notifications/src/grpc/server/convert.rs [hunk 2]
```diff
@@ -244,3 +244,30 @@ impl From<proto::DeepLinkAction> for notification_event::DeepLinkAction {
...
+                        .ok()
+                });
+
+                let dl = notification_event::DeepLink { screen, action };
+                notification_event::Action::OpenDeepLink(dl)
+            }
+            Some(proto::action::Data::ExternalUrl(url)) => {
+                notification_event::Action::OpenExternalUrl(url)
+            }
+            None => unreachable!("action data should always be set"),
```
Comment (thevaibhav-dixit) [inline core/notifications/src/grpc/server/convert.rs:270]: is there a better way to handle this ?

Comment (bodymindarts) [reply core/notifications/src/grpc/server/convert.rs:270]: You should use tryfrom

File: core/notifications/src/notification_event/marketing_notification_triggered.rs [hunk 3]
```diff
@@ -13,6 +13,7 @@ pub struct MarketingNotificationTriggered {
     pub should_add_to_history: bool,
     pub should_add_to_bulletin: bool,
     pub deep_link: Option<DeepLink>,
+    pub action: Option<Action>,
```
Comment (UncleSamtoshi) [inline core/notifications/src/notification_event/marketing_notification_triggered.rs:16]: @bodymindarts will this still properly deserialize payloads from the database that didn't have the action?

Comment (bodymindarts) [reply core/notifications/src/notification_event/marketing_notification_triggered.rs:16]: yes if you add `#[serde(default)]` - then non existent `action` attributes will deserialise to `None`

Comment (UncleSamtoshi) [reply core/notifications/src/notification_event/marketing_notification_triggered.rs:16]: Do you add it specifically to the `action` field or to the top level `MarketingNotificationTriggered` struct?

Comment (bodymindarts) [reply core/notifications/src/notification_event/marketing_notification_triggered.rs:16]: Just the field that was added.

File: core/notifications/src/notification_event/marketing_notification_triggered.rs [hunk 4]
```diff
@@ -12,7 +12,7 @@ pub struct MarketingNotificationTriggered {
     pub should_send_push: bool,
     pub should_add_to_history: bool,
     pub should_add_to_bulletin: bool,
-    pub deep_link: Option<DeepLink>,
+    pub deep_link: Option<DeepLink>, // This is no longer used but kept for deserialization of old events
```
Comment (thevaibhav-dixit) [inline core/notifications/src/notification_event/marketing_notification_triggered.rs:15]: can use ```#[serde(skip)]```
https://serde.rs/field-attrs.html#skip

Comment (bodymindarts) [reply core/notifications/src/notification_event/marketing_notification_triggered.rs:15]: Can just delete - serde will ignore superfluous attributes by default.

Comment (UncleSamtoshi) [reply core/notifications/src/notification_event/marketing_notification_triggered.rs]: Removed it

**Review by bodymindarts — APPROVED** [review]

**Review by sandipndev — APPROVED** [review]

**Pull request merged**